### PR TITLE
Add bluetooth stack

### PR DESCRIPTION
**Pull request by soypat**
Still a work in progress. Based on [dirbaio's excellent work on embassy-rs](https://github.com/embassy-rs/embassy/pull/2865) (is practically a direct, line-by-line port). Took me but a couple hours!

Need to figure out how HCI packets are differentiated from Ethernet packets, could not find it in the PR, yet.

Heads up to stakeholders @scottfeldman, @deadprogram. 

Summary of exported API added:

```go
func (d *Device) ReadHCI(b []byte) (int,error)
func (d *Device) WriteHCI(b []byte) (int,error)
func (d *Device) BufferedHCI() int
func DefaultBluetoothConfig() Config
func DefaultWifiBluetoothConfig() Config
```

### CODE REVIEW

File: bluetooth.go [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+				return err
+			}
+			time.Sleep(time.Millisecond) // TODO: is this sleep needed?
+		}
+	}
+	return nil
+}
+
+func (d *Device) hci_set_read_handler(fn func(b []byte) error) {
+
```
Comment (deadprogram) [inline bluetooth.go:157]: Is this supposed to be called from `RecvHCIHandle()` or was it left in by accident?

Comment (soypat) [reply bluetooth.go:157]: Ah, that was left in by accident. Thanks for catching it!

**Comment by soypat** [comment]
@deadprogram Interface matches rust version, should be ready to test!

**Comment by deadprogram** [comment]
Thanks for the update @soypat 

Now the error I get back is this:

```shell
unable to get reader from CYW43439 device: need to enable bluetooth in Init to use HCI interface
```

I am calling `Init()` in my test branch, unless I am doing it wrong somehow?

https://github.com/tinygo-org/bluetooth/pull/271/files#diff-bf4991e7a562e01fba5f8769cfae757a9df794f5ed14242000dc2beaa4df7a73R38-R44

**Comment by soypat** [comment]
Could you try running the [bluetooth example](https://github.com/soypat/cyw43439/blob/55b66a73813e03cb1e75def5f74f914185fb835c/examples/bluetooth/main.go) in the latest version of this branch to test if it is an issue with your setup (versioning/compiler/hardware)?

**Comment by deadprogram** [comment]
```shell
$ tinygo flash -size short -target pico -monitor ./examples/bluetooth/
   code    data     bss |   flash     ram
 401196   16164    3344 |  417360   19508
Connected to /dev/ttyACM0. Press Ctrl-C to exit.
panic: writeHCI:cyw: bt wake timeout
```

```shell
$ tinygo version
tinygo version 0.32.0-dev-30c4df16 linux/amd64 (using go version go1.22.0 and LLVM version 17.0.1)
```

**Comment by deadprogram** [comment]
With the latest commit in this branch:

```shell
$ tinygo flash -size=short -target=pico -tags=cyw43439 -monitor ./examples/scanner
   code    data     bss |   flash     ram
 421368   16248    3352 |  437616   19600
Connected to /dev/ttyACM0. Press Ctrl-C to exit.
Initializing CYW43439 device
Error initializing CYW43439 device cyw bt init failed: cyw: bt ready timeout
panic: failed to enable BLE stack: cyw bt init failed: cyw: bt ready timeout
```

File: bluetooth.go [hunk 2]
```diff
@@ -335,6 +336,18 @@ func (d *Device) hci_advance_read_ringbuf(n uint32) error {
...
 }
 
+func (d *Device) hci_ringbuf_debug() (h2btio_bt2hio [4]uint32) {
+	buf8 := u32AsU8(h2btio_bt2hio[:])
+	d.bp_read(d.btaddr+whd.BTSDIO_OFFSET_HOST2BT_IN, buf8)
+	d.trace("hci_ringbuf_debug",
+		slog.Uint64("h2bt_in", uint64(h2btio_bt2hio[0])),
+		slog.Uint64("h2bt_out", uint64(h2btio_bt2hio[1])),
+		slog.Uint64("bt2h_in", uint64(h2btio_bt2hio[2])),
+		slog.Uint64("h2bt_out", uint64(h2btio_bt2hio[3])),
```
Comment (harveysanders) [inline bluetooth.go:346]: Did you mean `bt2h_out` here?
```suggestion
		slog.Uint64("bt2h_out", uint64(h2btio_bt2hio[3])),
```

**Comment by soypat** [comment]
@deadprogram Is this ready for merge?

File: bluetooth.go [hunk 3]
```diff
@@ -0,0 +1,589 @@
...
+func (d *deviceHCI) Buffered() int {
+	return d.dev.BufferedHCI()
+}
+
+func (d *deviceHCI) Read(b []byte) (int, error) {
+	return d.dev.ReadHCI(b)
+}
+
+func (d *deviceHCI) Write(b []byte) (int, error) {
+	// if len(b)%4 != 0 && len(b) < len(d.buf4) {
```
Comment (deadprogram) [inline bluetooth.go:37]: I thought I removed these commented lines?

File: bluetooth.go [hunk 4]
```diff
@@ -0,0 +1,589 @@
...
+		const chunksize = 64 // Is writing in 64 byte chunks needed?
+		numChunks := alignedDataBufferIdx/64 + b2u32(alignedDataBufferIdx%64 != 0)
+		for i := uint32(0); i < numChunks; i++ {
+			offset := i * chunksize
+			end := (i + 1) * chunksize
+			if end > alignedDataBufferIdx {
+				end = alignedDataBufferIdx
+			}
+			chunk := bufferToWrite[offset:end]
+			// d.trace("chunk-write", slog.Uint64("idx", uint64(i)), slog.Uint64("offset", uint64(offset)), slog.Uint64("end", uint64(end)), slog.Uint64("len", uint64(end-offset)))
```
Comment (deadprogram) [inline bluetooth.go:222]: remove this?

**Comment by deadprogram** [comment]
@soypat I made a few nits, after which I think it can be squashed/merged.

**Comment by soypat** [comment]
@deadprogram :warning: pushed a breaking change, see commit 30a39e9. Also applied your nits and expanded the returned interface.

**Comment by deadprogram** [comment]
Checking now...

**Comment by deadprogram** [comment]
Works like a charm @soypat merge away!

**Comment by soypat** [comment]
Done! Thank you @deadprogram !

**Pull request merged**